File: 77_combinations/combine.py
class Solution(object):
    def combine(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: List[List[int]]
        """
        result = []

        def dfs(stack, num, k):
            if k == 0:
                result.append(stack)
                return

            for i in reversed(xrange(1, num + 1)):
                # Since k > num, we don't have enough num to fill --> return
                if k > i: break
                dfs(stack + [i], i - 1, k - 1)

        dfs([], n, k)
        return result

import time
sol = Solution()
n = 20
k = 16
start = time.time()
print sol.combine(n, k)
print time.time() - start

File: 77_combinations/combine.cpp
class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int> > res;
        dfs(res, k, n, vector<int>(), 0);
        return res;
    }
private:
    void dfs(vector<vector<int> > &res, int k, int n, vector<int> v, int begin) {
        if(k == v.size()) {
            res.push_back(v);
            return;
        }
        for(int i = begin; i < n; i++) {
            v.push_back(i + 1);
            dfs(res, k, n, v, i + 1);
            v.pop_back();
        }
    }
};